### PR TITLE
Created method for slicing ROI into tiles, for use inside `AsyncEffectRenderer`

### PR DESCRIPTION
**Pull request by Lehonti**
This should get us a step closer towards handling parallelism inside the effects themselves, as discussed in #808

### CODE REVIEW

File: Pinta.Core/Classes/AsyncEffectRenderer.cs [hunk 1]
```diff
@@ -44,24 +45,27 @@ internal abstract class AsyncEffectRenderer
...
-			if (tileWidth < 0) throw new ArgumentOutOfRangeException (nameof (tileWidth), "Cannot be negative");
-			if (tileHeight < 0) throw new ArgumentOutOfRangeException (nameof (tileHeight), "Cannot be negative");
+			if (renderBounds.Width < 0) throw new ArgumentException ("Width cannot be negative", nameof (renderBounds));
+			if (renderBounds.Height < 0) throw new ArgumentException ("Cannot be negative", nameof (renderBounds));
 			if (updateMilliseconds <= 0) throw new ArgumentOutOfRangeException (nameof (updateMilliseconds), "Strictly positive value expected");
 			if (threadCount < 1) throw new ArgumentOutOfRangeException (nameof (threadCount), "Invalid number of threads");
-			TileWidth = tileWidth;
-			TileHeight = tileHeight;
+
+			// If the effect isn't tileable, there is a single tile for the entire render bounds.
```
Comment (cameronwhite) [inline Pinta.Core/Classes/AsyncEffectRenderer.cs:65]: This comment doesn't really make sense here - maybe it should be lower down around the call to VerticalSliceup()?

File: Pinta.Core/Classes/AsyncEffectRenderer.cs [hunk 2]
```diff
@@ -244,11 +247,10 @@ void Render (int renderId, int threadId)
 	void RenderTile (int renderId, int threadId, int tileIndex)
 	{
 		Exception? exception = null;
-		var bounds = new RectangleI ();
+		RectangleI bounds = new ();
 
 		try {
-
-			bounds = GetTileBounds (tileIndex);
+			bounds = target_tiles[tileIndex]; // Get tile bounds
```
Comment (cameronwhite) [inline Pinta.Core/Classes/AsyncEffectRenderer.cs:253]: This doesn't really need to be in the try/catch block anymore

**Comment by Lehonti** [comment]
Thank you for the feedback @cameronwhite, I have addressed your points

File: Pinta.Core/Extensions/OtherExtensions.cs [hunk 3]
```diff
@@ -68,12 +68,25 @@ public static ReadOnlyCollection<T> ToReadOnlyCollection<T> (this ImmutableList<
...
+		internal ImmutableBackedReadOnlyCollection (ImmutableList<T> list)
+			: base (list)
+		{ }
+
+		internal ImmutableBackedReadOnlyCollection (ImmutableArray<T> array)
+			: base (array)
+		{ }
+	}
+
+	public static IEnumerable<RectangleI> VerticalSliceup (this RectangleI original)
```
[the remark this answers is not in the document]
Comment (Lehonti) [reply Pinta.Core/Extensions/OtherExtensions.cs]: It is now named `ToRows`

**Pull request merged**